pyro/contrib/gp/parameterized.py

```diff
@@ -0,0 +1,262 @@
+
+        # remove the constraint and its unconstrained parameter
+        self.set_constraint(name, constraints.real)
+
+        self._priors[name] = prior
+        self.autoguide(name, dist.Delta)
+
+    def autoguide(self, name, dist_constructor):
+        """
+        Sets an autoguide for a parameter with name ``name`` (mimic the behavior
+        of module :mod:`pyro.contrib.autoguide`).
+
+        ..note:: `dist_constructor` should be one of
+            :class:`~pyro.distributions.Delta`,
+            :class:`~pyro.distributions.Normal`, and
+            :class:`~pyro.distributions.MultivariateNormal`. More distribution
+            constructor will be supported in the future if needed.
+
+        :param str name: Name of the parameter.
+        :param dist_constructor: A
+            :class:`~pyro.distributions.distribution.Distribution` constructor.
+        """
+        if name not in self._priors:
+            raise ValueError("There is no prior for parameter: {}".format(name))
+
+        if dist_constructor not in [dist.Delta, dist.Normal, dist.MultivariateNormal]:
+            raise ValueError("Currently, only support autoguide for Delta, Normal, "
+                             "and MultivariateNormal distributions.")
+
+        if name in self._guides:
+            # delete previous guide's parameters
+            dist_args = self._guides[name][1]
+            for arg in dist_args:
+                arg_name = "{}_{}".format(name, arg)
+                if arg_name in self._constraints:
+                    # delete its unconstrained parameter
+                    self.set_constraint(arg_name, constraints.real)
+                delattr(self, arg_name)
+
+        # TODO: create a new argument `autoguide_args` to store other args for other
+        # constructors. For example, in LowRankMVN, we need argument `rank`.
+        p = self._buffers[name]
+        if dist_constructor is dist.Delta:
+            p_map = Parameter(p.detach())
+            self.register_parameter("{}_map".format(name), p_map)
+            self.set_constraint("{}_map".format(name), self._priors[name].support)
+            dist_args = {"map"}
+        elif dist_constructor is dist.Normal:
+            loc = Parameter(biject_to(self._priors[name].support).inv(p).detach())
+            scale = Parameter(loc.new_ones(loc.shape))
+            self.register_parameter("{}_loc".format(name), loc)
+            self.register_parameter("{}_scale".format(name), scale)
+            dist_args = {"loc", "scale"}
+        elif dist_constructor is dist.MultivariateNormal:
+            loc = Parameter(biject_to(self._priors[name].support).inv(p).detach())
+            identity = eye_like(loc, loc.size(-1))
+            scale_tril = Parameter(identity.repeat(loc.shape[:-1] + (1, 1)))
+            self.register_parameter("{}_loc".format(name), loc)
+            self.register_parameter("{}_scale_tril".format(name), scale_tril)
+            dist_args = {"loc", "scale_tril"}
```

Comment:
Can we also check
```py
else:
    raise NotImplementedError("unsupported distribution type: {}".format(dist_constructor))
```

Reply:
There already has a check for it from above. But I agree that there should be something to put here (because I used `elif`). Do you have any suggestion?

Reply:
```py
else:
    raise NotImplementedError
```